File: frontend/src/types/discovery.ts
export type EntityTypeName = 'character' | 'relationship' | 'world_bible';
export type DedupActionType = 'new' | 'exact_merge' | 'fuzzy_merge' | 'enrich';
export type JobStatusType = 'pending' | 'chunking' | 'extracting' | 'deduplicating' | 'staging' | 'completed' | 'failed';

export interface CascadeScanRequest {
  project_id: string;
  manuscript_text: string;
  chunk_size?: number;
  include_types?: EntityTypeName[];
}

export interface CascadeJobResponse {
  job_id: string;
  status: JobStatusType;
  phase: JobStatusType;
  chunk_index: number | null;
  total_chunks: number | null;
  stage_id: string | null;
  error: string | null;
}

export interface StagedEntity {
  entity_id: string;
  entity_type: EntityTypeName;
  entity_json: Record<string, unknown>;
  confidence: number;
  source_excerpt: string | null;
  approved: boolean;
  dedup_action: DedupActionType;
}

export interface StagedEntitiesResponse {
  stage_id: string;
  project_id: string;
  characters: StagedEntity[];
  relationships: StagedEntity[];
  world_bible: StagedEntity[];
}

export interface CascadeApplyResponse {
  characters_added: number;
  relationships_added: number;
  world_bible_added: number;
  characters_enriched: number;
}

export interface EntityApprovalUpdate {
  entity_id: string;
  approved: boolean;
}

File: app/schemas/discovery.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

EntityTypeName = Literal["character", "relationship", "world_bible"]
DedupActionType = Literal["new", "exact_merge", "fuzzy_merge", "enrich"]
JobStatusType = Literal["pending", "chunking", "extracting", "deduplicating", "staging", "completed", "failed"]


class CascadeScanRequest(BaseModel):
    project_id: str = Field(..., min_length=1)
    manuscript_text: str = Field(..., min_length=50)
    chunk_size: int = Field(default=8000, ge=1000, le=50000)
    include_types: list[EntityTypeName] = Field(
        default=["character", "relationship", "world_bible"],
    )


class CascadeJobResponse(BaseModel):
    job_id: str
    status: JobStatusType
    phase: JobStatusType
    chunk_index: int | None = None
    total_chunks: int | None = None
    stage_id: str | None = None
    error: str | None = None


class StagedEntity(BaseModel):
    entity_id: str
    entity_type: EntityTypeName
    entity_json: dict[str, object]
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    source_excerpt: str | None = None
    approved: bool = False
    dedup_action: DedupActionType


class EntityApprovalUpdate(BaseModel):
    entity_id: str
    approved: bool


class CascadeApplyResponse(BaseModel):
    characters_added: int = 0
    relationships_added: int = 0
    world_bible_added: int = 0
    characters_enriched: int = 0


class StagedEntitiesResponse(BaseModel):
    stage_id: str
    project_id: str
    characters: list[StagedEntity] = []
    relationships: list[StagedEntity] = []
    world_bible: list[StagedEntity] = []


class CascadeUndoResponse(BaseModel):
    stage_id: str
    entities_reverted: int
